sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Helpers/MethodSignatureHelper.cs

```diff
@@ -19,24 +19,28 @@
  */
 
 using System;
+using System.Linq;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.VisualBasic.Syntax;
 
 namespace SonarAnalyzer.Helpers
 {
     public static class MethodSignatureHelper
     {
-        public static Func<MethodSignature, bool> IsSame<TSymbolType>(SimpleNameSyntax identifierName, SemanticModel semanticModel)
+        public static bool IsMatch<TSymbolType>(SimpleNameSyntax identifierName, SemanticModel semanticModel,
+            Lazy<TSymbolType> symbolFetcher, MethodSignature[] methods)
             where TSymbolType : class, ISymbol
         {
+            if (identifierName == null)
+            {
+                return false;
+            }
+
             var identifierText = identifierName.Identifier.ValueText;
-            var identifierSymbol = new Lazy<TSymbolType>(() => semanticModel.GetSymbolInfo(identifierName).Symbol as TSymbolType);
-            // This function will be called multiple times for each tracked MethodSignature,
-            // hence we cache as much as possible before.
-            return (methodSignature) =>
-                identifierText == methodSignature.Name &&
-                identifierSymbol.Value != null &&
-                identifierSymbol.Value.ContainingType.Is(methodSignature.ContainingType);
+            return methods.Any(m =>
+                identifierText == m.Name &&
+                symbolFetcher.Value != null &&
+                symbolFetcher.Value.ContainingType.Is(m.ContainingType));
```

Comment:
I cannot believe I am saying that :), but I feel this LINQ would be probably better if written with a `foreach` (or even `for`) loop:
```
foreach (var m in methods)
{
  if (identifierText != m.Name)
    continue;
  if (symbolFetcher.Value == null)
    break; // No need to continue looping if the symbol is null
  if (symbolFetcher.Value.ContainingType.Is(m.ContainingType))
    return true;
}
return false;
```
Again, this looks like a micro-optimization, but could be something that we could change after we have more rules implemented.